### ISSUE TITLE
Turning arrays into lists

### ISSUE THREAD

**Issue by lionel-** [hadley/purrr#66]
To allow purrr's functions to talk to arrays, they need be transformed into lists.

`plyr::alply()` is a flexible tool that makes it is easy to turn arrays into lists, but it can be slow and hard to read. I think it makes sense to offer faster and more expressive functions in purrr for the most common coercions.

I think the most common tasks are flattening dimensions into a list, and enlisting dimensions hierarchically:
```r
flatten_margin <- function(array, margin = NULL) {
  if (is.null(dim(array))) {
    dim(array) <- length(array)
  }
  if (is.null(margin)) {
    margin <- seq_along(dim(array))
  }
  apply(array, margin, list) %>% flatten()
}

enlist_margin <- function(array, margin = NULL) {
  if (is.null(margin)) {
    margin <- seq_along(dim(array))
  }
  if (length(margin) > 1) {
    new_margin <- ifelse(margin[-1] > margin[[1]], margin[-1] - 1, margin[-1])
    apply(array, margin[[1]], . %>% enlist_margin(new_margin))
  } else {
    flatten_margin(array, margin)
  }
}
```

These functions are faster than the corresponding `alply()` operations.
```r
x <- array(seq_len(1e6), c(1000, 100, 10))
microbenchmark::microbenchmark(
  plyr = x %>% plyr::alply(2, .fun = . %>% plyr::alply(2)),
  purrr = x %>% enlist_margin(2:3)
)
```
```
neval: 100
   expr median         unit
1  plyr    322 milliseconds
2 purrr     66 milliseconds
```

And scale much better:
```r
x2 <- array(seq_len(1e6), c(10, 100, 1000))
microbenchmark::microbenchmark(
  plyr = x2 %>% plyr::alply(2, .fun = . %>% plyr::alply(2)),
  purrr = x2 %>% enlist_margin(2:3)
)
```
```
neval: 100
   expr median         unit
1  plyr  11479 milliseconds
2 purrr    241 milliseconds
```

They could be rewritten in Rcpp in the future for an additional performance boost.

**Comment by lionel-** [comment]
Here they are in action:

```r
x <- array(1:12, c(2, 2, 3))

flatten_margin(x) %>% str()
## List of 12
##  $ : int 1
##  $ : int 2
##  $ : int 3
##  $ : int 4
##  $ : int 5
##  $ : int 6
##  $ : int 7
##  $ : int 8
##  $ : int 9
##  $ : int 10
##  $ : int 11
##  $ : int 12
```

```r
flatten_margin(x, 1) %>% str()
## List of 2
##  $ : int [1:2, 1:3] 1 3 5 7 9 11
##  $ : int [1:2, 1:3] 2 4 6 8 10 12
```

```r
flatten_margin(x, c(1, 3) %>% str()
## List of 6
##  $ : int [1:2] 1 3
##  $ : int [1:2] 2 4
##  $ : int [1:2] 5 7
##  $ : int [1:2] 6 8
##  $ : int [1:2] 9 11
##  $ : int [1:2] 10 12
```

```r
enlist_margin(x) %>% str()
## List of 2
##  $ :List of 2
##   ..$ :List of 3
##   .. ..$ : int 1
##   .. ..$ : int 5
##   .. ..$ : int 9
##   ..$ :List of 3
##   .. ..$ : int 3
##   .. ..$ : int 7
##   .. ..$ : int 11
##  $ :List of 2
##   ..$ :List of 3
##   .. ..$ : int 2
##   .. ..$ : int 6
##   .. ..$ : int 10
##   ..$ :List of 3
##   .. ..$ : int 4
##   .. ..$ : int 8
##   .. ..$ : int 12
```

```r
enlist_margin(x, c(3, 1)) %>% str()
## List of 3
##  $ :List of 2
##   ..$ : int [1:2] 1 3
##   ..$ : int [1:2] 2 4
##  $ :List of 2
##   ..$ : int [1:2] 5 7
##   ..$ : int [1:2] 6 8
##  $ :List of 2
##   ..$ : int [1:2] 9 11
##   ..$ : int [1:2] 10 12
```

Should I send a PR?

**Comment by hadley** [comment]
Nice! Please send a PR.

**Comment by lionel-** [comment]
Implemented by #68
